### DISCUSSION TITLE
is there a way to add a custom plugin not packaged in nixpkgs?

### DISCUSSION THREAD

**Discussion by conscious-puppet** [nix-community/nixvim#592] [Q&A] [answered]
i want to add a plugin by giving its github url. i could do this by using `fetchFromGithub` and then, but is there any helper function provided by nixvim?

**Comment by gaetan-puleo** [comment]
Here an exemple in my config (it's a flake)

```nix
{ 
  
 ... # inputs

  # plugins
  inputs.nvim-window-picker-src = {
    url = "github:s1n7ax/nvim-window-picker";
    flake = false;
  };

  outputs = {
    self,
    nixpkgs,
    nixvim,
    flake-utils,
    nvim-window-picker-src
  }:
    flake-utils.lib.eachDefaultSystem (system: let
      pkgs = nixpkgs.legacyPackages."${system}";
      nvim = nixvim.legacyPackages."${system}".makeNixvim {
        ... # config here
 
        extraPlugins = [
          (pkgs.vimUtils.buildVimPlugin {
          name = "nvim-window-picker";
          src = nvim-window-picker-src;
          })
        ];
      };
    in {
     ...
    });
}
```

**Comment by ptonner** [comment]
is there a way to use this approach when using `makeNixvimWithModule`? It doesn't seem like the module gets access to anything other than `pkgs` so I'm not sure how to provide these extra inputs to the module

**Reply by ptonner** [reply]
to partially answer my own question, it appears that I can use overlays as shown here: https://github.com/traxys/nvim-flake/blob/529cf14d62717b2b5ef45669f148294e8912564c/flake.nix#L248

Is this the right way to do it?

**Reply by gaetan-puleo** [reply]
You can use an overlay yes. On my side I am using a flake to build my config, this is why I can add inputs

**Reply by ptonner** [reply, marked as answer]
right, I'm using flakes as well but my call to `makeNixvimWithModule` uses a module file that won't have a reference to the flake `inputs` directly, just `pkgs`. This is how `makeNixvimWithModule` is written, so I think there is no way around it without overlays. But I wanted to double check my assumptions in case there's an alternative.

To make things more clear, here's my basic setup. A top-level flake:
```nix
{ 
    inputs = {
        #....
        extra-plugin = {
          url = "github:extra/plugin";
          flake = false;
        };
    }
    outputs = { nixpkgs, flake-utils, nixvim, ... }@inputs:
    flake-utils.lib.eachDefaultSystem (system:
      let
        pkgs = import nixpkgs {
          inherit system;
        };
        nvim = nixvim.legacyPackages.${system}.makeNixvimWithModule {
          inherit pkgs;
          module = ./nix;
        };
      in {
        packages = {
          inherit nvim;
          default = nvim;
        };
      });
}
```
And all the config happens in `./nix`, where there won't be a direct reference to `extra-plugin` available.

To be clear, I've got something working already with overlays, but just want to double check there isn't an alternative I'm unaware of

**Comment by GaetanLepage** [comment]
You can also very easily upstream a vim package.
If you need me to do it, feel free to file an issue :)
It is also possible to add plugins to nixvim of course !

**Reply by conscious-puppet** [reply]
thanks for that. it's a personal plugin that i am not so confident about upstreaming. 
i am new to nix, so i am trying to understand the various suggestions people have made.

**Reply by GaetanLepage** [reply]
Ok ! In this case, using `vimUtils.buildVimPlugin` is the best approach !

**Comment by GaetanLepage** [comment]
I am marking this discussion as solved :)
